pages/blog/internals-of-nova-part-2.md

```diff
@@ -0,0 +1,286 @@
+
+To sidestep this issue, local memory caches have been added to the processors. These caches are both closer to the CPU, reducing access latency from distance, and more importantly use a [faster, volatile memory technology](https://en.wikipedia.org/wiki/Static_random-access_memory) than main memory. Fetching data from the main memory to the local caches is slow, but once the data is there then re-fetching it is massively faster. A program will often reuse the same memory multiple times over, and that memory will then be accessed from caches, avoiding the need to load the data from main memory.
+
+To give concrete numbers, a main memory load takes approximately 200 nanoseconds. A local cache access takes between 1 and 40 nanoseconds, depending on the cache level. That is up to two orders of magnitude better! And the effect is generally additive: Loading data that is then used to load more data pays the memory latency twice. Often a program will also have very deep memory load trees, and at worst every branch in the tree pays the 200 nanosecond price: This is what crashes and burns program performance in this day and age. It is not a slow CPU bringing you low, nor a slow hard drive or a lack of memory. It is often not an undersized GPU that grinds a game's performance to a halt: It is memory loading, or "memory stalls", that ruins your day.
+
+Data-oriented design is all about solving or at least mitigating this issue. To do so, we must know our data and use that knowledge to minimize memory usage which always improves cache usage, and we must take into account the CPU caches when desigining our software. The important thing here is that a CPU loads data into cache not byte-by-byte but in cache lines. This is usually 64 bytes at a time, though recent CPUs have bumped this number up to 128 bytes.
+
+```rs
+#[repr(align(64))]
+struct CacheLine([u8; 64]);
+```
+
+When we load a piece of data we load a full cache line into the CPU, no matter how small a part of it we actually want to use. Loading this one cache line from main memory takes 200 nanoseconds. Now, if we only use a single byte (or worse, one bit!) of that data then the vast majority of the data loaded is wasted. What we want is to ensure that as much of the cache line gets used, especially if we are in a tight loop.
+
```

Comment:
**suggestion**: Maybe you could add another heading or horizontal rule here to split this section into two parts: "what is DOD" and "the 'applied' example".